src/shell/shell_misc.cpp

```diff
@@ -72,331 +74,255 @@ void DOS_Shell::ShowPrompt(void) {
+	line[CMD_MAXLINE - 1] = '\0';
+
+	const auto* const dos_section = dynamic_cast<Section_prop*>(
+	        control->GetSection("dos"));
+	assert(dos_section != nullptr);
+	const std::string_view expand_shell_variable_pref = dos_section->Get_string(
+	        "expand_shell_variable");
+	if ((expand_shell_variable_pref == "auto" && dos.version.major >= 7) ||
+	    expand_shell_variable_pref == "true") {
+		ProcessCmdLineEnvVarStitution(line);
+	}
+}
+
+std::string DOS_Shell::ReadCommand()
+{
+	std::vector<std::string> history_clone = history;
+	history_clone.emplace_back("");
+	auto history_index = history_clone.size() - 1;
 
-	reset_str(line);
+	std::vector<std::string> completion = {};
+	auto completion_index               = completion.size();
 
-	std::list<std::string>::iterator it_history = l_history.begin(), it_completion = l_completion.begin();
+	std::string command   = {};
+	auto cursor_position  = command.size();
+	auto completion_start = command.size();
+
+	uint8_t data        = 0;
+	uint16_t byte_count = 1;
 
 	Int10Manager int10_manager;
 
-	while (size && !shutdown_requested) {
+	while (!shutdown_requested) {
 		bool viewing_tab_completions = false;
-		dos.echo=false;
-		while(!DOS_ReadFile(input_handle,&c,&n)) {
-			uint16_t dummy;
+
+		while (!DOS_ReadFile(input_handle, &data, &byte_count)) {
+			uint16_t dummy = 1;
 			DOS_CloseFile(input_handle);
-			DOS_OpenFile("con",2,&dummy);
-			LOG(LOG_MISC,LOG_ERROR)("Reopening the input handle. This is a bug!");
+			DOS_OpenFile("con", 2, &dummy);
+			LOG(LOG_MISC, LOG_ERROR)
+			("Reopening the input handle. This is a bug!");
 		}
-		if (!n) {
-			size=0;			//Kill the while loop
-			continue;
+
+		if (byte_count == 0) {
+			break;
 		}
-		switch (c) {
-		case 0x00: /* Extended Keys */
-			{
-				DOS_ReadFile(input_handle,&c,&n);
-				switch (c) {
-
-				case 0x3d: /* F3 */
-					if (!l_history.size()) break;
-					it_history = l_history.begin();
-					if (it_history != l_history.end() && it_history->length() > str_len) {
-						const char *reader = &(it_history->c_str())[str_len];
-						while ((c = *reader++)) {
-							line[str_index ++] = c;
-						}
-						str_len = str_index = it_history->length();
-						size = CMD_MAXLINE - str_index - 2;
-						terminate_str_at(line, str_len);
-					}
-					break;
 
-				case 0x4B: /* Left */
-					if (str_index) {
-						str_index --;
-					}
+		constexpr uint8_t ExtendedKey = 0x00;
+		constexpr uint8_t Escape      = 0x1B;
+		switch (data) {
+		case ExtendedKey: {
+			DOS_ReadFile(input_handle, &data, &byte_count);
+
+			constexpr uint8_t ShiftPlusTab = 0x0F;
+			constexpr uint8_t F3           = 0x3D;
+			constexpr uint8_t Home         = 0x47;
+			constexpr uint8_t Up           = 0x48;
+			constexpr uint8_t Left         = 0x4B;
+			constexpr uint8_t Right        = 0x4D;
+			constexpr uint8_t End          = 0x4F;
+			constexpr uint8_t Down         = 0x50;
+			constexpr uint8_t Delete       = 0x53;
+			switch (data) {
```

Comment:
Don't we already have all DOS keycodes defined somewhere? If not, could you perhaphs start a file with these common scancodes?